Sources/NIO/CircularBuffer.swift

```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    ///
+    /// *O(n)* where _n_ is the length of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange<C>(_ subrange: RangeType<Int>, with newElements: C) where C : Collection, E == C.Element {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex, "Subrange out of bounds")
+
+        let lowerBound = bufferIndex(ofIndex: subrange.lowerBound)
+        let upperBound = bufferIndex(ofIndex: subrange.upperBound == self.endIndex ? subrange.upperBound - 1 : subrange.upperBound)
+
+        if subrange.count == newElements.count {
+            for (index, element) in zip(subrange.indices, newElements) {
+                self.buffer[self.bufferIndex(ofIndex: index)] = element
+            }
+        } else if subrange.count == self.count && newElements.isEmpty {
+            self.removeSubrange(subrange)
+        } else {
+            var newBuffer: ContiguousArray<E?> = []
+            let capacityDelta = (newElements.count - subrange.count)
+            let newCapacity = Int(UInt32(self.buffer.count + capacityDelta).nextPowerOf2())
+            newBuffer.reserveCapacity(self.buffer.count)
+            var index = self.headIdx
+            while index != self.tailIdx {
+                if index == lowerBound {
+                    for element in newElements {
+                        newBuffer.append(element)
+                    }
+                    while index != upperBound {
+                        index = self.bufferIndex(after: index)
+                    }
+                } else {
+                    newBuffer.append(self.buffer[index])
+                    index = self.bufferIndex(after: index)
+                }
+            }
+
+            self.tailIdx = newBuffer.count
+            let repetitionCount = self.buffer.count - newBuffer.count
+            if repetitionCount > 0 {
+                newBuffer.append(contentsOf: repeatElement(nil, count: repetitionCount))
+            }
+            self.headIdx = 0
+            self.buffer = newBuffer
+        }
+    }
+
+    public mutating func replaceSubrange<C, R>(_ subrange: R, with newElements: C) where C : Collection, R : RangeExpression, E == C.Element, Int == R.Bound {
+        let range = subrange.relative(to: self)
+        self.replaceSubrange(RangeType<Int>(range), with: newElements)
+    }
+
+    public mutating func removeSubrange(_ bounds: Range<Int>) {
+        if bounds.count == 1 {
+            _ = remove(at: bounds.lowerBound)
+        } else if bounds.count == self.count {
+            var newBuffer: ContiguousArray<E?> = []
+            newBuffer.reserveCapacity(self.buffer.count)
+            newBuffer.append(contentsOf: repeatElement(nil, count: self.count))
+            self.headIdx = 0
+            self.tailIdx = 0
+            self.buffer = newBuffer
```

Comment:
You can do this whole block with `self = .init(initialRingCapacity: self.buffer.count)`